pkg/timelock/timelock.go

```diff
@@ -128,166 +128,256 @@ func NewTimelockWorker(nodeURL, timelockAddress, callProxyAddress, privateKey st
 		return err
 	}
+	defer sub.Unsubscribe()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case err := <-sub.Err():
+				// Check if the error is not nil, because sub.Unsubscribe will
+				// signal the channel sub.Err() to close it, leading to false nil errors.
+				if err != nil {
+					tw.logger.Warn().Msgf("subscription error: %s", err.Error())
+					SetReadyStatus(HealthStatusError)
+					sub.Unsubscribe()
+
+					success := false
+					for try := range maxSubRetries {
+						tw.logger.Warn().Msgf("trying to re-create subscription: %v/%v retry.", try+1, maxSubRetries)
+						sub, err = tw.ethClient.SubscribeFilterLogs(ctx, query, logCh)
+						if err == nil {
+							tw.logger.Info().Msg("subscription successfully recreated.")
+							SetReadyStatus(HealthStatusOK)
+							success = true
+							break
+						}
+
+						time.Sleep(time.Second * time.Duration(try))
+					}
+
+					if !success {
+						tw.logger.Error().Msg("failed to recreate subscription after retries: shutting down timelock-worker.")
+						wg.Done()
+						return
+					}
+				}
+
+			case <-ctx.Done():
+				tw.logger.Debug().Msgf("shutting down subscription")
+				SetReadyStatus(HealthStatusError)
+				return
+			}
+		}
+	}()
 
-	// Read events by FilterLogs. This method calls eth_getLogs under the hood.
+	return nil
+}
+
+func (tw *Worker) retrieveHistoricalLogs(ctx context.Context) (chan types.Log, error) {
+	query := tw.setupFilterQuery()
+	logCh := make(chan types.Log)
+
+	// FilterLogs starts filtering the logs at fromBlock, gathering the historical data.
+	// This is needed to guarantee that all the events are gathered, even in scenarios where the worker crashes.
 	filter, err := tw.ethClient.FilterLogs(ctx, query)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
+	// Process incoming historical logs in a separate goroutine.
+	wg.Add(1)
 	go func() {
-		for _, l := range filter {
-			logCh <- l
+		defer wg.Done()
+		for _, log := range filter {
+			select {
+			case logCh <- log:
+				tw.logger.Debug().Msgf("processing historical log: %v\n", log)
+			case <-ctx.Done():
+				tw.logger.Debug().Msg("shutting down the retrieval of old logs in incomplete status")
+				return
+			}
 		}
+		tw.logger.Debug().Msg("retrieved correctly all historical events")
+		close(logCh)
 	}()
 
-	// Setting readyStatus here because we want to make sure subscription is up.
-	tw.logger.Info().Msgf("Initial subscription complete")
-	SetReadyStatus(HealthStatusOK)
+	return logCh, nil
+}
 
-	// This is the goroutine watching over the subscription.
-	// We want wg.Done() to cancel the whole execution, so don't add more than 1 to wg.
-	// Also, when receiving an event that creates an error, skip the event and
-	// continue processing the rest, as an external operator can cancel the faulty event.
-	loop := true
+func (tw *Worker) processLogs(ctx context.Context, oldLog, newLog chan types.Log, stopCh chan string, cancel context.CancelFunc) {
```

Comment:
```suggestion
func (tw *Worker) processLogs(stopCh <-chan string, oldLog, newLog chan types.Log) {
```

Reply:
I really don't think you should pass context and stopCh here.  Instead each of the sub go routines just gets a context and you have the `handleOSSignal` routine that's only job is to log the error.  In that case, `handleOSSignal` should just return the child context and cancel func made from `signal.NotifyContext`

Reply:
`stopCh` seems redundant if you use `signal.NotfiyContext`.  if using a child context of `signal.NotifyContext` everywhere then `ctx.Err()` will have the signal error if any.  So you can always get that error from any go routine.